src/main.ts

```diff
@@ -218,25 +220,104 @@ function _checkOptionValidity(
 
 function _makeDefaultContext<T>(
   cloudFunction: CloudFunction<T>,
-  options: ContextOptions
+  options: ContextOptions,
+  triggerData?: T
 ): EventContext {
   let eventContextOptions = options as EventContextOptions;
+  const eventResource = cloudFunction.__trigger.eventTrigger?.resource;
+  const eventType = cloudFunction.__trigger.eventTrigger?.eventType;
+
+  const optionsParams = eventContextOptions.params ?? {};
+  let triggerParams = {};
+  if (eventResource && eventType && triggerData) {
+    if (eventType.startsWith('google.firebase.database.ref.')) {
+      let data: database.DataSnapshot;
+      if (eventType.endsWith('.write')) {
+        // Triggered with change
+        if (!(triggerData instanceof Change)) {
+          throw new Error('Must be triggered by database change');
+        }
+        data = triggerData.before;
+      } else {
+        data = triggerData as any;
+      }
+      triggerParams = _extractDatabaseParams(eventResource, data);
+    } else if (eventType.startsWith('google.firestore.document.')) {
+      let data: firestore.DocumentSnapshot;
+      if (eventType.endsWith('.write')) {
+        // Triggered with change
+        if (!(triggerData instanceof Change)) {
+          throw new Error('Must be triggered by firestore document change');
+        }
+        data = triggerData.before;
+      } else {
+        data = triggerData as any;
+      }
+      triggerParams = _extractFirestoreDocumentParams(eventResource, data);
```

Comment:
I don't _really_ like these bits but not really sure there's a better way?

I was thinking to check `data instanceof database.DataSnapshot` (or `firestore.DocumentSnapshot`) and throwing an `Error` but then if you call the wrapped function with an actual fetched `database.DataSnapshot` (using the admin SDK) it would fail that check which would be annoying!